### PR TITLE
Fix HHL().construct_circuit() properly encoding the eigenvalues of certain matrixes

### PR DESCRIPTION
**Pull request by jeppevinkel**
### Summary
This fixes an issue where HHL().construct_circuit() doesn't always provide the correct number of qubits to properly represent the eigenvalues due to an issue with the rounding.

### Details and comments
The fix is implemented with a small change in the math used to compute the size of the eigenvalue evaluation register to properly indicute the amount of bits needed to accurately represent a given number.

### CODE REVIEW

**Comment by anedumla** [comment]
@woodsp-ibm does it work now? Usually I need to comment on the issues to be able to get assigned to them.

**Comment by anedumla** [comment]
@jeppevinkel  could you give an example for which the old code fails? In theory `n_l` should depend on the accuracy but in practice this would lead to too many qubits and for most examples the simulation would take too long or would not be possible at all.

**Comment by jeppevinkel** [comment]
Matrixes with certain condition numbers fail such as this one with a condition number of 3.1244, [82, 34; 34, 58]. It seems to be an issue near the points where my modified solution would increase the number of qubits generally.

**Comment by anedumla** [comment]
Yes, you were right, the problem is that with the previous calculation `delta` was `0.0` because we need that 
image.
Thanks a lot for your input!

**Comment by anedumla** [comment]
@woodsp-ibm  I cannot approve this PR, but it looks good to me.

File: releasenotes/notes/fix-hhl_construct_circuit-nl-size-03cbfba9ed50a57a.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    Fixes issue where HHL().construct_circuit() don't always return a correct circuit for some matrixes due to not having enough qubits to represent the eigenvalues.
```
Comment (woodsp-ibm) [inline releasenotes/notes/fix-hhl_construct_circuit-nl-size-03cbfba9ed50a57a.yaml:4]:
```suggestion
    Fixes an issue where HHL().construct_circuit() didn't always return a correct circuit. Issue arose for some matrices due to not allocating enough qubits to represent the eigenvalues.
```

**Comment by woodsp-ibm** [comment]
A made a small suggestion for the text in the release note. Is a unit test possible here? we normally have a unit test added to ensure the cases/bugs that presumably were not caught by unit testing are now at least checked going forwards.

**Comment by jeppevinkel** [comment]
A unit test should be possible yeah. Once I get some time I'll see if I can set one up.

**Comment by Cryoris** [comment]
@jeppevinkel are you still pursuing this? 🙂

**Comment by jeppevinkel** [comment]
Yeah sorry I'm currently very busy with University exams, so I haven't had the time to implement the unit test yet.

**Comment by jeppevinkel** [comment]
I'm sorry for how long it took, but university has been stressful lately. I've now written a proposal for a test in [test/python/algorithms/test_linear_solvers.py](https://github.com/Qiskit/qiskit-terra/pull/7402/files#diff-0f3fae09e957c2436bd09d40ffdc2cf769040ac38b657768d5e984183e85f17d) that fails without this change in the calculated size of `nl`

**Comment by jeppevinkel** [comment]
> Thanks for fixing this issue you noticed. Hope the university exams went well.

Yeah I'm just honored to be able to contribute. I'm aspiring to be a programmer, so I'm just happy to be able to contribute to an open source project.

**Pull request merged**